### PR TITLE
Bind the rest of the CMFormatDescription types

### PR DESCRIPTION
**Pull request by timrisi**
This fixes comments from the original PR and adds some tests.  Original PR at https://github.com/xamarin/xamarin-macios/pull/262

### CODE REVIEW

**Comment by spouliot** [comment]
any reason this could not have been continued on the original PR ? we're losing the history of the comments :(

**Comment by timrisi** [comment]
I rebased onto the latest master, so I couldn't just push to the same branch.  If there's a way I can keep it as part of the original PR still I'm happy to do so.

**Comment by spouliot** [comment]
Even after a rebase (merging master) you should still be able to push to the same branch and have the original PR, e.g. I did it twice to solve conflicts in https://github.com/xamarin/xamarin-macios/pull/1026

**Comment by rolfbjarne** [comment]
You can force push to branches in your own fork.

**Comment by timrisi** [comment]
The original PR isn't on my fork, it's a branch on xamarin/xamarin-macios

**Comment by rolfbjarne** [comment]
Yeah, then you can't rebase. So always use your own fork 😄

**Comment by timrisi** [comment]
Yah, I do now.  This PR is from my fork :).  The original PR for coremediametadata was right after switching to open source to get it moved over and before I'd forked myself.

**Comment by rolfbjarne** [comment]
The introspection-ios tests failed:

> [FAIL] ApiCMAttachmentTest.CheckFailAttachments : System.InvalidOperationException : Could not create the new instance for type CMMuxedFormatDescription.

**Comment by timrisi** [comment]
Build failure is unrelated (build timeout in watchos fsharp tests)

File: src/CoreMedia/CMFormatDescription.cs [hunk 1]
```diff
@@ -292,20 +326,17 @@ static CMFormatDescription Create (CMMediaType type, IntPtr handle, bool owns)
...
-		[DllImport (Constants.CoreMediaLibrary)]
-		extern static /* AudioFormatListItem* */ IntPtr CMAudioFormatDescriptionGetRichestDecodableFormat (/* CMAudioFormatDescriptionRef */ IntPtr desc);
-
-		public AudioFormat AudioRichestDecodableFormat {
+		public AudioFormat? AudioRichestDecodableFormat {
 			get {
 				unsafe {
 					var ret = (AudioFormat *) CMAudioFormatDescriptionGetRichestDecodableFormat (handle);
 					if (ret == null)
-						return new AudioFormat ();
```
Comment (chamons) [inline src/CoreMedia/CMFormatDescription.cs]: Is this a behavior change users might care about?

File: src/CoreMedia/CMFormatDescription.cs [hunk 2]
```diff
@@ -374,7 +490,153 @@ internal CMAudioFormatDescription (IntPtr handle, bool owns)
...
+		{
+		}
+
+		public CMAudioFormatDescription (ref AudioStreamBasicDescription asbd, AudioChannelLayout layout, uint magicCookieSize, IntPtr magicCookie, NSDictionary extensions)
+		{
+			IntPtr ptr = IntPtr.Zero;
+			int size = 0;
+			if (layout != null) {
+				size = Marshal.SizeOf (layout.NativeStruct);
+				ptr = Marshal.AllocHGlobal (size);
```
Comment (chamons) [inline src/CoreMedia/CMFormatDescription.cs]: Who is deallocing this later?

File: src/CoreMedia/CMFormatDescription.cs [hunk 3]
```diff
@@ -540,4 +839,217 @@ public bool VideoMatchesImageBuffer (CVImageBuffer imageBuffer)
...
+			var error = CMMetadataFormatDescriptionCreateWithMetadataSpecifications (IntPtr.Zero, metadataType, metadataSpecifications != null ? metadataSpecifications.Handle : IntPtr.Zero, out handle);
+			if (error != CMFormatDescriptionError.None)
+				throw new ArgumentException (error.ToString ());
+
+			return new CMMetadataFormatDescription (handle, true);
+		}
+
+		[Mac (10,10)]
+		[DllImport (Constants.CoreMediaLibrary)]
+		static extern CMFormatDescriptionError CMMetadataFormatDescriptionCreateWithMetadataFormatDescriptionAndMetadataSpecifications (
```
Comment (chamons) [inline src/CoreMedia/CMFormatDescription.cs]: What a mouthful...

File: tests/apitest/src/CoreMedia/CMFormatDescriptionTests.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+using NUnit.Framework;
+
+namespace Xamarin.Mac.Tests
+{
+	[TestFixture]
+	public class CMFormatDescriptionTests
+	{
+		public CMFormatDescriptionTests ()
+		{
+			//var videoDesc = CMFormatDescription.Create (CMMediaType.Video, 61766331, out error);
```
Comment (chamons) [inline tests/apitest/src/CoreMedia/CMFormatDescriptionTests.cs]: I'm assuming these are not meant to be commented out?

File: src/CoreMedia/CMFormatDescription.cs (hunk 2 above)
Comment (timrisi) [inline src/CoreMedia/CMFormatDescription.cs]: Updated to call Marshal.FreeHGlobal after it's used and no longer needed.

File: tests/apitest/src/CoreMedia/CMFormatDescriptionTests.cs (hunk 4 above)
Comment (timrisi) [inline tests/apitest/src/CoreMedia/CMFormatDescriptionTests.cs]: Removed the whole constructor.  The commented code is stuff I copy pasted over from an XM app where I was figuring out how the code worked and forgot to delete it after putting it into tests.

**Comment by chamons** [comment]
@spouliot Could you review and let @timrisi know what else (if anything) needs done on the ancient PR?

**Comment by timrisi** [comment]
@spouliot Hey, have you had a chacne to review this again yet?

**Comment by chamons** [comment]
@timrisi I forgot. We are blocked on this waiting for a good sample from QA to test (beyond your auto test).

**Comment by spouliot** [comment]
@chamons it seems you have not approved all of your requested changes

**Comment by spouliot** [comment]
@olegoid we're in need of a sample for those API, thanks!

**Comment by olegoid** [comment]
@spouliot thanks for reminding me. This request is tracked on our sample building board:
https://trello.com/c/tnECVnkr

**Comment by rolfbjarne** [comment]
Looks like we have bug #[32636](https://bugzilla.xamarin.com/show_bug.cgi?id=32636) requesting this API.

File: src/CoreMedia/CMFormatDescription.cs (hunk 2 above)
Comment (Therzok) [reply src/CoreMedia/CMFormatDescription.cs]: Can we not just pass the structure by ref instead of manually allocating marshal memory?

**Comment by mandel-macaque** [comment]
Closing as PR was abandoned and create an issue to follow up: https://github.com/xamarin/xamarin-macios/issues/7485

**Pull request closed without merging**